src/VecSim/spaces/computer/computer.h

```diff
@@ -0,0 +1,357 @@
+template <typename DistType, typename DistFuncType>
+ProcessedBlobs
+IndexComputerBasic<DistType, DistFuncType>::preprocess(const void *original_blob,
+                                                       size_t processed_bytes_count) const {
+    auto storage_blob = preprocessForStorage(original_blob, processed_bytes_count);
+    auto query_blob = preprocessQuery(original_blob, processed_bytes_count);
+    return ProcessedBlobs(std::move(storage_blob), std::move(query_blob));
+}
+
+template <typename DistType, typename DistFuncType>
+MemoryUtils::unique_blob IndexComputerBasic<DistType, DistFuncType>::preprocessForStorage(
+    const void *original_blob, size_t processed_bytes_count) const {
+    return MemoryUtils::unique_blob(const_cast<void *>(original_blob),
+                                    MemoryUtils::dummyFreeAllocation);
+}
+
+template <typename DistType, typename DistFuncType>
+MemoryUtils::unique_blob
+IndexComputerBasic<DistType, DistFuncType>::preprocessQuery(const void *original_blob,
+                                                            size_t processed_bytes_count) const {
+    return maybeCopyToAlignedMem(original_blob, processed_bytes_count);
+}
+
+template <typename DistType, typename DistFuncType>
+MemoryUtils::unique_blob
+IndexComputerBasic<DistType, DistFuncType>::maybeCopyToAlignedMem(const void *original_blob,
+                                                                  size_t blob_bytes_count) const {
+    if (this->alignment) {
+        if ((this->alignment && (uintptr_t)original_blob % this->alignment)) {
+            auto aligned_mem =
+                this->allocator->allocate_force_aligned(blob_bytes_count, this->alignment);
+            memcpy(aligned_mem, original_blob, blob_bytes_count);
+            return MemoryUtils::unique_blob(
+                aligned_mem, [this](void *ptr) { this->allocator->free_allocation(ptr); });
+        }
+    }
+
+    // Returning a unique_ptr with a no-op deleter
+    return MemoryUtils::unique_blob(const_cast<void *>(original_blob),
+                                    MemoryUtils::dummyFreeAllocation);
+}
+
+/* ======================= IndexComputerExtended ======================= */
+
+template <typename DistType, typename DistFuncType>
+int IndexComputerExtended<DistType, DistFuncType>::addPreprocessor(
+    PreprocessorAbstract *preprocessor) {
+    for (size_t i = 0; i < n_preprocessors; i++) {
+        if (preprocessors[i] == nullptr) {
+            if (preprocessor->hasQueryPreprocessor()) {
+                flags |= Flags::QUERY;
+            }
+
+            if (preprocessor->hasStoragePreprocessor()) {
+                flags |= Flags::STORAGE;
+            }
+
+            preprocessors[i] = preprocessor;
+            return i + 1 >= n_preprocessors ? 0 : i + 1;
+        }
+    }
+    return -1;
+}
+
+template <typename DistType, typename DistFuncType>
+ProcessedBlobs
+IndexComputerExtended<DistType, DistFuncType>::preprocess(const void *original_blob,
```

Comment:
I don't get why this function is so complex.
Why isn't it a single loop over the preprocessors, and calling it on one or both blobs depends on its flags?
Why do we need to support an incomplete state where the preprocessors have not been initialized yet?